src/fairseq2/models/wav2vec2/model.py

```diff
@@ -113,7 +113,9 @@ def __init__(
         self.logit_temp = logit_temp
         self.diversity_loss_weight = diversity_loss_weight
 
-    def forward(self, batch: SequenceBatch) -> Wav2Vec2Output:
+    def forward(
+        self, batch: SequenceBatch, return_features: bool = False
+    ) -> Wav2Vec2Output | Tuple[Wav2Vec2Output, torch.Tensor]:
```

Comment:
This makes sense as a standalone feature (i.e. we should be able to return intermediate features as part of the forward call). But for @mshvartsman 's usecase (of finetuning a pretrained wav2vec2 model), this approach, while works, is inefficient since we are done extra computations. I would suggest the follwing:

1. We add a `extract_feature` method to extract just the features (which solves @mshvartsman 's usecase efficiently) and
2. We add the support for returning the intermediate features to `Wav2Vec2Output` i.e. `return_features` flag is propogated to (and handled in) `self.quantize_and_contrast`

Reply:
The extra compute is because we `quantize_and_contrast` even if we want features only? Equivalent solution would be to change `return_features` flag to `features_only` flag and then return `Wav2Vec2Output | torch.Tensor`, thoughts? Agree that `extract_features` should be its own method though.

Reply:
You mean extra compute since we `quantize_and_contrast` even if we only want the features? Fair point, but I think we have a choice here: 
1. If I want logits or both features and logits to be the cheap option, current behavior is fine. Features-only (e.g. fine-tune workflow) requires extra compute. 
2. If we want logits or features but not both to be cheap, we can have `features_only` argument and return `Wav2Vec2Output | torch.Tensor`, but then having both (e.g. to record both val loss and representation metrics during SSL, which is also a valid use case) requires running the encoder twice. 
3. If we want all options to be cheap, we need to return not just features but `padding_mask, targets, temporal_mask` from `extract_features` so we can do whatever we want with the features and then pass everything down to `quantize_and_contrast`. But then I think downstream code will often need to call `features, _, _, _ = extract_features(...)` which is an awkward call unless we want it to also have a `features_only` argument. 

Thoughts? My lean is [3] but all of them feel a bit awkward.

Reply:
The output of the `extract_features` function can be a structured object, just like the output of the `forward` call](https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/models/wav2vec2/model.py#L116). So the `forward` call can also use the `extract_features` function. Regarding `extract_features` having a `features_only` argument, I think it is reasonable.

Reply:
I think with a structured output of `extract_features` we don't need to worry about `features_only` since the compute cost doesn't change and it's easy to ignore the extra stuff.